packages/core/src/queue/aggregate.js

```diff
@@ -0,0 +1,133 @@
+import { parse as parseLink } from 'multiformats/link'
+
+import { connect } from '../database/index.js'
+import {
+  SQLSTATE_UNIQUE_VALUE_CONSTRAINT_ERROR_CODE,
+  DEFAULT_LIMIT
+} from '../database/constants.js'
+import {
+  DatabaseOperationError,
+  DatabaseValueToUpdateAlreadyTakenError,
+  DatabaseValueToUpdateAlreadyTakenErrorName
+} from '../database/errors.js'
+
+export const TABLE_NAME = 'aggregate'
+export const INCLUSION_TABLE_NAME = 'inclusion'
+export const VIEW_NAME = 'aggregate_queue'
+
+/**
+ * 
+ * @param {import('../types.js').DatabaseConnect} conf
+ * @returns {import('../types.js').AggregateQueue}
+ */
+export function createAggregateQueue (conf) {
+  const dbClient = connect(conf)
+
+  return {
+    put: async (aggregateItems) => {
+      const items = aggregateItems.map(aggregateItem => ({
+        link: `${aggregateItem.link}`,
+        size: aggregateItem.size,
+        pieces: aggregateItem.pieces
+      }))
+
+      try {
+        // Transaction
+        await dbClient.transaction().execute(async trx => {
+          // Insert to aggregate table
+          const insertOps = await trx
+            .insertInto(TABLE_NAME)
+            .values(items.map(item => ({
+              link: item.link,
+              size: item.size
+            })))
+            // NOOP if item is already in table
+            .onConflict(oc => oc
+              .column('link')
+              .doNothing()
+            )
+            .execute()
+
+          // Update inclusion table to point to aggregates
+          const updateOpsPerAggregate = await Promise.all(items.map(item => trx
+            .updateTable(INCLUSION_TABLE_NAME)
+            .set({
+              aggregate: item.link.toString()
+            })
+            .where('aggregate', 'is', null)
+            .where('piece', 'in', item.pieces.map(i => i.toString()))
+            .execute()
+          ))
```

Comment:
I don't know how to do all that with kysely but we could do conditional updates without rollbacks as follows

```sql
WITH pieces AS (
  SELECT * FROM (VALUES ('a'), ('b'), ('c')) AS t(piece)
),
used_pieces AS (
  SELECT COUNT(*) AS count
  FROM inclusion
  WHERE piece IN (SELECT piece FROM pieces)
  AND aggregate IS NOT NULL
),
used_aggregate AS (
  SELECT COUNT(*) AS count
  FROM aggregate
  WHERE link = 'aggregate_cid'
),
include AS (
  UPDATE inclusion
  SET aggregate = 'aggregate_cid'
  WHERE piece IN (SELECT piece FROM pieces)
  AND aggregate IS NULL
  AND (SELECT count FROM used_pieces) = 0
  AND (SELECT count FROM used_aggregate) = 0
)
SELECT * FROM include;
```

Advantage of this approach is that we'll end up adding only aggregates that can be added and ignore ones that could not be added.

Reply:
That would be almost the default behavior if I did `onConflict`. However, I do not appreciate to have `Put` failing partially and return success. We changed to single items to avoid all this